### PR TITLE
Support execute dataframe with compose in both local and distributed mode

### LINKED ISSUE #175: Support run dataframe in local cluster and distributed mode
**Is your feature request related to a problem? Please describe.**

Currently, Mars DataFrame can only run on local thread-based executor. 

**Describe the solution you'd like**

We need to let the local cluster and distributed mode support DataFrame either.

**Additional context**

This is following the PR #173 .

### LINKED ISSUE #408: [BUG] Execute dataframe failed due to the graph compose
**Describe the bug**

When executing dataframe with `compose=True`, it will fail when compose the graph.

**To Reproduce**
```Python
In [1]: import pandas as pd                                                                         

In [2]: import numpy as np                                                                          

In [3]: from mars.dataframe.expressions.datasource.dataframe import from_pandas 
   ...: from mars.dataframe.expressions.arithmetic import add                                       

In [4]: data1 = pd.DataFrame(np.random.rand(10, 10), index=[0, 10, 2, 3, 4, 5, 6, 7, 8, 9], 
   ...:                              columns=[4, 1, 3, 2, 10, 5, 9, 8, 6, 7])                       

In [5]: data2 = pd.DataFrame(np.random.rand(10, 10), index=[11, 1, 2, 5, 7, 6, 8, 9, 10, 3], 
   ...:                              columns=[5, 9, 12, 3, 11, 10, 6, 4, 1, 2])                     

In [6]: df1 = from_pandas(data1, chunk_size=5)                                                      

In [7]: df2 = from_pandas(data2, chunk_size=6)                                                      

In [8]: df3 = add(df1, df2)                                                                         

In [9]: df3.execute()                                                                               
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-9-cc354b1c63e3> in <module>
----> 1 df3.execute()

~/Documents/mars_dev/mars/mars/core.py in execute(self, session, **kw)
    485         if session is None:
    486             session = Session.default_or_local()
--> 487         return session.run(self, **kw)
    488 
    489     def fetch(self, session=None, **kw):

~/Documents/mars_dev/mars/mars/session.py in run(self, *tileables, **kw)
    113         tileables = tuple(mt.tensor(t) if not isinstance(t, (Entity, Base)) else t
    114                           for t in tileables)
--> 115         result = self._sess.run(*tileables, **kw)
    116 
    117         for t in tileables:

~/Documents/mars_dev/mars/mars/session.py in run(self, *tileables, **kw)
     54         if 'n_parallel' not in kw:
     55             kw['n_parallel'] = cpu_count()
---> 56         res = self._executor.execute_tileables(tileables, **kw)
     57         return res
     58 

~/Documents/mars_dev/mars/mars/utils.py in _wrapped(*args, **kwargs)
    351         try:
    352             _kernel_mode.eager = False
--> 353             return func(*args, **kwargs)
    354         finally:
    355             _kernel_mode.eager = None

~/Documents/mars_dev/mars/mars/executor.py in execute_tileables(self, tileables, fetch, n_parallel, n_thread, print_progress, mock, compose)
    521 
    522             tileable.build_graph(graph=graph, tiled=True, compose=compose,
--> 523                                  executed_keys=list(self._chunk_result.keys()))
    524 
    525         self.execute_graph(graph, result_keys, n_parallel=n_parallel or n_thread,

~/Documents/mars_dev/mars/mars/core.py in build_graph(self, graph, cls, tiled, compose, executed_keys)
    448                           if c not in visited])
    449         if tiled and compose:
--> 450             graph.compose(keys=keys)
    451 
    452         if not tiled and any(not n.is_coarse() for n in graph):

~/Documents/mars_dev/mars/mars/graph.pyx in mars.graph.DirectedGraph.compose()

~/Documents/mars_dev/mars/mars/fuse.pyx in mars.fuse.Fusion.compose()

~/Documents/mars_dev/mars/mars/fuse.pyx in mars.fuse.Fusion._compose_graph()

AttributeError: 'DataFrameChunkData' object has no attribute 'dtype'
```

**Expected behavior**
Return result as execute with `compose=False`.
```Python
In [10]: df3.execute(compose=False)                                                                 
Out[10]: 
          1         2         3         4         5         6   7   8         9         10  11  12
0        NaN       NaN       NaN       NaN       NaN       NaN NaN NaN       NaN       NaN NaN NaN
1        NaN       NaN       NaN       NaN       NaN       NaN NaN NaN       NaN       NaN NaN NaN
2   1.076695  1.204041  1.060153  0.509292  1.339779  1.342729 NaN NaN  0.715889  1.400814 NaN NaN
3   0.965725  0.592176  1.282727  1.196449  0.529224  0.831382 NaN NaN  0.744456  1.250852 NaN NaN
4        NaN       NaN       NaN       NaN       NaN       NaN NaN NaN       NaN       NaN NaN NaN
5   1.072791  1.489320  1.295868  0.947721  1.711967  0.906912 NaN NaN  1.102863  0.516068 NaN NaN
6   0.461877  0.993711  1.520775  0.767697  1.428225  1.687759 NaN NaN  0.816748  0.623632 NaN NaN
7   1.051975  0.906304  0.819521  1.701052  1.478845  1.645575 NaN NaN  0.482702  0.998527 NaN NaN
8   1.052598  0.339266  1.381926  1.048586  1.314938  0.597436 NaN NaN  0.342243  0.278099 NaN NaN
9   1.294781  0.865981  0.773909  1.477216  1.114498  1.207362 NaN NaN  0.773577  0.514921 NaN NaN
10  0.364438  1.238505  1.031440  0.396337  0.593371  0.413832 NaN NaN  0.607879  0.148480 NaN NaN
11       NaN       NaN       NaN       NaN       NaN       NaN NaN NaN       NaN       NaN NaN NaN

```

### LINKED ISSUE #409: [BUG] DataFrame's columns can't be serialized
**Describe the bug**

Dataframe can't be serialized when its columns's value type is `IndexValue.Int64Index`.

**To Reproduce**

```Python
In [1]: import pandas as pd                                                                         

In [2]: import numpy as np                                                                          

In [3]: from mars.dataframe.expressions.datasource.dataframe import from_pandas 
   ...: from mars.dataframe.expressions.arithmetic import add                                       

In [4]: data1 = pd.DataFrame(np.random.rand(10, 10), index=[0, 10, 2, 3, 4, 5, 6, 7, 8, 9], 
   ...:                              columns=[4, 1, 3, 2, 10, 5, 9, 8, 6, 7])                       

In [5]: df1 = from_pandas(data1, chunk_size=5)                                                      

In [6]: df1.build_graph().to_json()                                                                 
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-74f9d74aaf91> in <module>
----> 1 df1.build_graph().to_json()

~/Documents/mars_dev/mars/mars/graph.pyx in mars.graph.DirectedGraph.to_json()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Serializable.to_json()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Serializable.serialize()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Provider.serialize_model()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Field.serialize()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Field.serialize()

~/Documents/mars_dev/mars/mars/serialize/jsonserializer.pyx in mars.serialize.jsonserializer.JsonSerializeProvider.serialize_field()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Serializable.serialize()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Provider.serialize_model()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Field.serialize()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Field.serialize()

~/Documents/mars_dev/mars/mars/serialize/jsonserializer.pyx in mars.serialize.jsonserializer.JsonSerializeProvider.serialize_field()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Serializable.serialize()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Provider.serialize_model()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Field.serialize()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Field.serialize()

~/Documents/mars_dev/mars/mars/serialize/jsonserializer.pyx in mars.serialize.jsonserializer.JsonSerializeProvider.serialize_field()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Serializable.serialize()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Provider.serialize_model()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Field.serialize()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Field.serialize()

~/Documents/mars_dev/mars/mars/serialize/jsonserializer.pyx in mars.serialize.jsonserializer.JsonSerializeProvider.serialize_field()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Serializable.serialize()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Provider.serialize_model()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Field.serialize()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Field.serialize()

~/Documents/mars_dev/mars/mars/serialize/jsonserializer.pyx in mars.serialize.jsonserializer.JsonSerializeProvider.serialize_field()

~/Documents/mars_dev/mars/mars/serialize/jsonserializer.pyx in mars.serialize.jsonserializer.JsonSerializeProvider._serialize_value()

~/Documents/mars_dev/mars/mars/serialize/jsonserializer.pyx in mars.serialize.jsonserializer.JsonSerializeProvider._serialize_typed_value()

~/Documents/mars_dev/mars/mars/serialize/jsonserializer.pyx in mars.serialize.jsonserializer.JsonSerializeProvider._serialize_list()

~/Documents/mars_dev/mars/mars/serialize/jsonserializer.pyx in mars.serialize.jsonserializer.JsonSerializeProvider._serialize_value()

~/Documents/mars_dev/mars/mars/serialize/jsonserializer.pyx in mars.serialize.jsonserializer.JsonSerializeProvider._serialize_untyped_value()

TypeError: Unknown type to serialize: <class 'numpy.int64'>

In [7]: list(df1.build_graph())[0].columns.to_json()                                                
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-8-d12b2c877a9d> in <module>
----> 1 list(df1.build_graph())[0].columns.to_json()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Serializable.to_json()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Serializable.serialize()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Provider.serialize_model()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Field.serialize()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Field.serialize()

~/Documents/mars_dev/mars/mars/serialize/jsonserializer.pyx in mars.serialize.jsonserializer.JsonSerializeProvider.serialize_field()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Serializable.serialize()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Provider.serialize_model()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Field.serialize()

~/Documents/mars_dev/mars/mars/serialize/core.pyx in mars.serialize.core.Field.serialize()

~/Documents/mars_dev/mars/mars/serialize/jsonserializer.pyx in mars.serialize.jsonserializer.JsonSerializeProvider.serialize_field()

~/Documents/mars_dev/mars/mars/serialize/jsonserializer.pyx in mars.serialize.jsonserializer.JsonSerializeProvider._serialize_value()

~/Documents/mars_dev/mars/mars/serialize/jsonserializer.pyx in mars.serialize.jsonserializer.JsonSerializeProvider._serialize_typed_value()

~/Documents/mars_dev/mars/mars/serialize/jsonserializer.pyx in mars.serialize.jsonserializer.JsonSerializeProvider._serialize_list()

~/Documents/mars_dev/mars/mars/serialize/jsonserializer.pyx in mars.serialize.jsonserializer.JsonSerializeProvider._serialize_value()

~/Documents/mars_dev/mars/mars/serialize/jsonserializer.pyx in mars.serialize.jsonserializer.JsonSerializeProvider._serialize_untyped_value()

TypeError: Unknown type to serialize: <class 'numpy.int64'>

```

### PR DESCRIPTION
**Pull request by NJUPole**
## What do these changes do?

Fix serialization of  DataFrame's columns and support execute DataFrame with `compose=True`.

## Related issue number

fixes #408 
fixes #409 
resolve #175

### CODE REVIEW

File: mars/dataframe/execution/arithmetic.py [hunk 1]
```diff
@@ -82,15 +83,15 @@ def _index_align_map(ctx, chunk):
 
 
 def _index_align_reduce(ctx, chunk):
-    input_idx_to_df = {inp.index: ctx[inp.key, ','.join(str(idx) for idx in chunk.index)]
-                       for inp in chunk.inputs[0].inputs}
-    row_idxes = sorted({idx[0] for idx in input_idx_to_df})
-    col_idxes = sorted({idx[1] for idx in input_idx_to_df})
+    input_keys = get_shuffle_reduce_inputs(chunk.inputs[0])
+    idxes_gen = itertools.product(range(chunk.op.row_source_size), range(chunk.op.col_source_size))
```
Comment (chineking) [inline mars/dataframe/execution/arithmetic.py]: it's indeed a little bizarre to  handle all chunk inputs' indexes in such a way, I recommend to add a `_to_fetch_idxes` to `FetchShuffle`, and `row_source_size` and `col_source_size` could not be required any more.

File: mars/tensor/execution/reshape.py [hunk 2]
```diff
@@ -109,16 +108,15 @@ def _reshape_map_estimate_size(ctx, chunk):
 
 
 def _reshape_reduce(ctx, chunk):
+    from ...utils import get_shuffle_reduce_inputs
```
Comment (chineking) [inline mars/tensor/execution/reshape.py]: Better to move the import outside.

**Comment by wjsi** [comment]
Will this completely solves #175 ?

File: mars/dataframe/core.py [hunk 3]
```diff
@@ -374,6 +374,7 @@ class DataFrameChunkData(ChunkData):
     _dtypes = SeriesField('dtypes')
     _index_value = ReferenceField('index_value', IndexValue)
     _columns_value = ReferenceField('columns_value', IndexValue)
+    _composed = ListField('composed', ValueType.reference('self'))
```
[the remark this answers is not in the document]
Comment (NJUPole) [reply mars/dataframe/core.py]: it will be resolved in #412

File: mars/core.py [hunk 4]
```diff
@@ -322,7 +322,7 @@ def params(self):
         p = {
```
Comment (chineking) [inline mars/core.py]: How about just return the tail node's `params`?

File: mars/serialize/jsonserializer.pyx [hunk 5]
```diff
@@ -443,7 +443,10 @@ cdef class JsonSerializeProvider(Provider):
                 if field.weak_ref:
                     val = val()
                 if val is not None:
-                    new_obj.append(val.serialize(self, dict()))
+                    if isinstance(val, Serializable):
```
Comment (chineking) [inline mars/serialize/jsonserializer.pyx]: Seems better to check if val is the reference type.

File: mars/serialize/pbserializer.pyx [hunk 6]
```diff
@@ -473,7 +473,10 @@ cdef class ProtobufSerializeProvider(Provider):
                         val = val()
                     it_obj = add()
                     if val is not None:
-                        val.serialize(self, obj=it_obj)
+                        if isinstance(val, Serializable):
```
Comment (chineking) [inline mars/serialize/pbserializer.pyx]: ditto.

File: mars/core.py (hunk 4 above)
Comment (NJUPole) [reply mars/core.py]: Done

**Pull request merged**